### PR TITLE
fix(llama): distinguish zephyr/tinyllama from phi chat template

### LINKED ISSUE #1271: Bug: TinyLlama-1.1B-Chat's e2e case produces empty output — first decode step predicts EOS immediately
### Affected area

Inference output or task execution

### TensorRT-Model-Connect version

296f3661 (upstream/main at time of testing)

### Installation method

Development container

### Model and revision

TinyLlama/TinyLlama-1.1B-Chat-v1.0 at `main`

### Environment

- OS: Ubuntu 24.04 (container)
- GPU: NVIDIA A40, 48GB VRAM
- Driver: 580.159.03 (CUDA forward-compatibility mode, CUDA 13.3 userspace)
- CUDA: 13.3
- TensorRT: 11.1.0.106
- Container image: nvcr.io/nvidia/tensorrt:26.07-py3, with torch==2.12.0+cu130 (matches Dockerfile.dev.x86-gpu / community GPU CI environment)

### Minimal reproducer

```bash
# Build the bundle
python3.12 -c "
from pathlib import Path
from tensorrt_model_connect import BuildRequest, build
build(BuildRequest(
    model_dir=Path('/path/to/TinyLlama-1.1B-Chat-v1.0'),
    output_path=Path('/tmp/tinyllama-1.1b.bundle'),
    family='llama', task='text_generation', precision='fp16',
    max_sequence_length=256,
))
"

# Run it through the native runtime
trtmc run /tmp/tinyllama-1.1b.bundle \
  --runtime-root <build-output-dir> \
  --prompt "What is the capital of France? Answer in one word." \
  --max-new-tokens 10 \
  --use-chat-template true
```

This is also directly reproducible through the repository's own existing e2e
harness, without any external reproducer:

```bash
TRTMC_BINARY=<path-to-trtmc> TRTMC_RUNTIME_ROOT=<build-output-dir> \
  pytest families/llama/tests/test_e2e.py --e2e-model=tinyllama-1.1b -s
```

using the manifest already checked into the repository at
`families/llama/tests/manifests/tinyllama-1.1b.json`.

### Observed behavior

The native runtime's first decode step (immediately after prefill) predicts
the EOS token, producing empty output:

```json
{"decode_ms":0.04743,"prefill_ms":67.912788,"setup_ms":0.00046,"text":"","token_ids":[2]}
```

The `families/llama/tests/test_e2e.py::test_e2e[tinyllama-1.1b]` case fails
its correctness assertion (`assert actual_text`) as a result.

### Expected behavior

A real HF reference generation (`transformers.AutoModelForCausalLM.generate()`,
same prompt, same chat template, greedy/`do_sample=False`) produces a
sensible, correct answer:

```
token_ids: [12415, 2410, 29901, 3681, 2]
text: "Capital: Paris"
```

The native engine's very first decode-step logits diverge sharply from the
reference at that step — this is not an EOS-matching issue (EOS=2 is the
correct id and correctly stops generation once predicted); the problem is
that EOS gets predicted immediately instead of real content.

### Suspected root cause

`families/llama/runtime/chat_templates.cpp`'s format-detection heuristic
(`llama_detect_chat_template_format`) substring-matches the raw Jinja
template for `<|user|>`/`<|assistant|>` role tags and classifies any template
containing them as **"Phi"** format. TinyLlama-1.1B-Chat's own chat template
also uses those role tags, but separates turns with its real EOS token
(`</s>`, id 2), not Phi's literal `<|end|>` turn-separator string that
`apply_phi()` injects. If TinyLlama's template is being misclassified as
Phi format, the rendered prompt fed to the engine would contain a
turn-separator token sequence the model was never trained on, plausibly
explaining an out-of-distribution first-step prediction. This is a
hypothesis based on reading the detection logic and is not proven with a
token-level trace of the actual rendered prompt — flagging it as a lead for
whoever investigates, not a confirmed diagnosis.

### Isolation notes

This is unrelated to any pending PR. Found incidentally while
regression-testing an unrelated `eos_token_id` fix (#1269) against the
ordinary single-EOS case. Reproduced identically by rebuilding with the
unmodified `families/llama/config.py`, `model.py`, and `runtime/plugin.cpp`
from `upstream/main`, confirming this predates and is unrelated to that PR.
`families/llama/runtime/chat_templates.cpp`, `pipeline.cpp`, and
`sampler.cpp` were not modified in that PR either.

### Relevant logs

```
[trtmc] KV cache rows=256 (bundle max=256)
[trtmc.prefill] tokens=13 launches=1 max_chunk=13
{"decode_ms":0.04743,"prefill_ms":67.912788,"setup_ms":0.00046,"text":"","token_ids":[2]}
[trtmc.engine_timing] label="llama prefill" execute_ms=56.055233 launches=1
[trtmc.engine_timing] label="llama decoder" execute_ms=12.112448 launches=1
```

### PR DESCRIPTION
**Pull request by dundysm**
## Background

Fixes #1271.

`families/llama/tests/manifests/tinyllama-1.1b.json` e2e fails because the native
runtime's first decode step predicts EOS immediately (`token_ids: [2]`, empty
text). HF reference greedy generation for the same prompt and chat template
returns a real answer (`token_ids: [12415, 2410, 29901, 3681, 2]`, text
`Capital: Paris`).

root cause (confirmed via template + detection/render comparison, not GPU
logits): `llama_detect_chat_template_format` treated any jinja containing
`<|user|>` / `<|assistant|>` as **phi**. TinyLlama-1.1B-Chat (and Zephyr) use
those role tags but end turns with `eos_token` (`</s>`), not Phi's literal
`<|end|>`. `apply_phi()` therefore injected `<|end|>` separators the model was
never trained on, which matches the observed out-of-distribution first-step EOS.

unrelated to open PR #1269 (eos_token_id list); that work does not touch
chat-template detection/rendering.

## Exit Criteria

- TinyLlama / Zephyr jinja (role tags + `eos_token`, no `<|end|>`) detects as
  `zephyr`, not `phi`.
- Phi jinja (role tags + `<|end|>`) still detects as `phi`.
- `llama_apply_chat_template("zephyr", ...)` renders
  `<|user|>\n{prompt}</s>\n<|assistant|>\n` (matches HF `apply_chat_template`
  with `trim_blocks`/`lstrip_blocks`).
- non-goal: does not change eos_token_id list handling (#1269).
- non-goal of this PR alone: full A40 GPU e2e (no local NVIDIA GPU here); ask
  maintainers/CI GPU to confirm the e2e harness.

## Implementation

in `families/llama/runtime/chat_templates.cpp`:

- require `<|end|>` (in addition to role tags) to classify **phi**.
- otherwise classify role-tag templates without `<|end|>` as **zephyr**.
- add `apply_zephyr()` that uses `</s>` turn ends instead of `<|end|>`.

regression coverage in `families/llama/tests/cpp/test_llama_chat_template.cpp`
(detect + apply + e2e prompt string; asserts no `<|end|>` in zephyr render).

affected: llama family chat-template detection/apply only. no public API, ABI,
or bundle format change (format string is internal; bundles still store the
same `chat_template.jinja` section).

### Change categories

- [x] Model or runtime behavior
- [ ] Public API
- [ ] ABI
- [ ] Bundle or artifact format
- [ ] Dependencies
- [ ] Documentation only
- [ ] CI or developer tooling

### Commands and Results

```
g++ -std=c++17 -Wall -Wextra -Wpedantic -I. \
  -o /tmp/test_llama_chat_template \
  families/llama/tests/cpp/test_llama_chat_template.cpp \
  families/llama/runtime/chat_templates.cpp
/tmp/test_llama_chat_template
# All llama chat_template tests passed.
```

also verified detect/apply on the real TinyLlama and Phi-3 jinja strings:

- TinyLlama tokenizer_config chat_template -> `zephyr`
- Phi-3 (`<|end|>` present) -> `phi`
- zephyr apply for the issue prompt:
  `<|user|>\nWhat is the capital of France? Answer in one word.</s>\n<|assistant|>\n`
- phi apply (broken path TinyLlama was hitting):
  `<|user|>\nWhat is the capital of France? Answer in one word.<|end|>\n<|assistant|>\n`

HF-equivalent jinja render (jinja2 `trim_blocks=True`, `lstrip_blocks=True`,
same as transformers) matches the zephyr apply string above.

### Hardware, Environment, and Revisions

- tested head: this branch (`fix/tinyllama-chat-template-format`)
- CPU-only box; **no NVIDIA GPU / no nvidia-smi**; cannot run A40 e2e or
  `trtmc run` / `pytest families/llama/tests/test_e2e.py --e2e-model=tinyllama-1.1b`
- TinyLlama template taken from `TinyLlama/TinyLlama-1.1B-Chat-v1.0` tokenizer_config
  on HF `main`; Phi comparison from `microsoft/Phi-3-mini-4k-instruct`

### Not Run / Remaining Gaps

- full GPU e2e for `tinyllama-1.1b` (build bundle + native vs HF reference) was
  **not** run locally. please confirm on community GPU CI / a machine with CUDA
  that `pytest families/llama/tests/test_e2e.py --e2e-model=tinyllama-1.1b`
  now produces non-empty text aligned with the HF reference cited in #1271.
- other llama e2e manifests not re-run (chat-template path for phi-tagged models
  unchanged aside from requiring `<|end|>` for phi classification).

## Contributor Self-Review

- [x] I have completed a self-review of this change.

## Notes For Future Readers

- suggested review order: detection change in `chat_templates.cpp`, then
  `apply_zephyr`, then new unit tests.
- if a future model uses `<|user|>` with a non-`</s>` eos, detection may need a
  richer signal; TinyLlama/Zephyr eos is `</s>` and matches this apply path.

### Risk level

- [x] Low
- [ ] Medium
- [ ] High

Risk rationale: surgical heuristic tightening plus one new format path; phi
behavior preserved when `<|end|>` is present; covered by non-GPU unit tests.

### CODE REVIEW

**Comment by dundysm** [comment]
@yifeif-nv @chaofengw-nv friendly ping for review when you have a moment; this is a small llama chat-template detection fix for #1271 (tinyllama misclassified as phi). no local gpu here so the unit tests are cpu-only; would appreciate a gpu ci / a40 pass on `pytest families/llama/tests/test_e2e.py --e2e-model=tinyllama-1.1b` if convenient.

**Comment by yifeif-nv** [comment]
> @yifeif-nv @chaofengw-nv friendly ping for review when you have a moment; this is a small llama chat-template detection fix for #1271 (tinyllama misclassified as phi). no local gpu here so the unit tests are cpu-only; would appreciate a gpu ci / a40 pass on `pytest families/llama/tests/test_e2e.py --e2e-model=tinyllama-1.1b` if convenient.

Hello, thanks for the contribution. I will take a look today and trigger the CI for you if it all looks good. Sorry I missed that PR earlier!

**Comment by yifeif-nv** [comment]
> @yifeif-nv @chaofengw-nv friendly ping for review when you have a moment; this is a small llama chat-template detection fix for #1271 (tinyllama misclassified as phi). no local gpu here so the unit tests are cpu-only; would appreciate a gpu ci / a40 pass on `pytest families/llama/tests/test_e2e.py --e2e-model=tinyllama-1.1b` if convenient.

Looks like a legitimate fix. I'm triggering the CI for you! Thanks for the contribution, and looking forward to future contributions as well!

**Comment by dundysm** [comment]
@yifeif-nv thanks for taking a look and kicking ci; really appreciate it.

pushed a small follow-up for the source-quality fail (`30c99c9`, clang-format on the zephyr/tinyllama string literal wrapping). community cpu is green on the new head now; still no local gpu here so would love the a40 / e2e pass when you get a chance. happy to iterate on anything else you spot.
